File: src/main/java/endytkn/randomEvents/goals/GroupFightGoal.java
package endytkn.randomEvents.goals;

import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.ai.goal.target.TargetGoal;
import net.minecraft.world.entity.ai.targeting.TargetingConditions;

import java.util.List;

public class GroupFightGoal extends TargetGoal {
    private final Mob mob;
    private final Class<? extends Mob> targetClass;

    public GroupFightGoal(Mob mob, Class<? extends Mob> targetClass) {
        super(mob, false);
        this.mob = mob;
        this.targetClass = targetClass;
    }

    @Override
    public boolean canUse() {
        TargetingConditions targetConditions = TargetingConditions.forCombat().ignoreLineOfSight();
        List<Mob> nearbyMobs = (List<Mob>) mob.level().getNearbyEntities(targetClass, targetConditions, mob, mob.getBoundingBox().inflate(10.0));

        String groupHate = mob.getPersistentData().getString("REGroupFightHates");

        for (Mob targetMob : nearbyMobs) {
            String targetGroup = targetMob.getPersistentData().getString("REGroupFightName");

            if (targetGroup.equals(groupHate)) {
                mob.setTarget(targetMob); // Set the target mob
                return true;
            }
        }
        return false;
    }

    @Override
    public void start() {
        LivingEntity target = mob.getTarget();
        if (target != null) {
            mob.setTarget(target);
        }
    }
}

File: src/main/kotlin/endytkn/randomEvents/baseEvents/GroupFight/GroupFight.goal.kt
package endytkn.randomEvents.baseEvents.GroupFight

import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.Mob
import net.minecraft.world.entity.ai.goal.target.TargetGoal
import net.minecraft.world.entity.ai.targeting.TargetingConditions

class GroupFightGoal(
    private val mob: Mob,
    private val targetClass: Class<out Mob>,
): TargetGoal(mob, false) {
    override fun canUse(): Boolean {
        val targetConditions = TargetingConditions.forCombat().ignoreLineOfSight()
        val nearbyMobs = mob.level().getNearbyEntities(targetClass, targetConditions, mob, mob.boundingBox.inflate(10.0))

        val groupHate = mob.persistentData.getString("REGroupFightHates")

        for (targetMob in nearbyMobs) {
            val targetGroup = targetMob.persistentData.getString("REGroupFightName")

            if (targetGroup == groupHate) {
                mob.target = targetMob as LivingEntity  // Set the target mob
                return true
            }
        }
        return false
    }

    override fun start() {
        mob.target?.let {
            mob.setTarget(it)
        }
    }
}